File: crypto/sha256.go
package crypto

import (
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"math"
	"math/big"
	"strconv"
)

// Diff1Target BigInt constant (0x00000000ffff0000000000000000000000000000000000000000000000000000)
var Diff1Target *big.Int

func init() {
	Diff1Target = new(big.Int)
	Diff1Target.SetString("00000000ffff0000000000000000000000000000000000000000000000000000", 16)
}

// Sha256d performs double SHA-256 (SHA-256(SHA-256(data)))
func Sha256d(data []byte) []byte {
	h1 := sha256.Sum256(data)
	h2 := sha256.Sum256(h1[:])
	return h2[:]
}

// ReverseBytes returns a new byte slice with reversed byte order
func ReverseBytes(src []byte) []byte {
	res := make([]byte, len(src))
	for i := 0; i < len(src); i++ {
		res[i] = src[len(src)-1-i]
	}
	return res
}

// NbitsToTarget converts compact nBits hex string to 256-bit BigInt target
func NbitsToTarget(nbitsHex string) *big.Int {
	nbits, err := strconv.ParseUint(nbitsHex, 16, 32)
	if err != nil {
		return big.NewInt(0)
	}
	exponent := uint(nbits >> 24)
	mantissa := big.NewInt(int64(nbits & 0x00ffffff))

	target := new(big.Int)
	if exponent <= 3 {
		target.Rsh(mantissa, 8*(3-exponent))
	} else {
		target.Lsh(mantissa, 8*(exponent-3))
	}
	return target
}

// NbitsToDifficulty converts nBits hex string to network difficulty
func NbitsToDifficulty(nbitsHex string) float64 {
	target := NbitsToTarget(nbitsHex)
	if target.Sign() == 0 {
		return 0
	}
	diff1Float := new(big.Float).SetInt(Diff1Target)
	targetFloat := new(big.Float).SetInt(target)
	resFloat := new(big.Float).Quo(diff1Float, targetFloat)
	res, _ := resFloat.Float64()
	return res
}

// DifficultyToTarget converts numeric difficulty to 256-bit target BigInt
func DifficultyToTarget(diff float64) *big.Int {
	if diff <= 0 {
		return new(big.Int).Set(Diff1Target)
	}
	diff1Float := new(big.Float).SetInt(Diff1Target)
	diffFloat := big.NewFloat(diff)
	targetFloat := new(big.Float).Quo(diff1Float, diffFloat)

	target := new(big.Int)
	targetFloat.Int(target)
	return target
}

// HashToDifficulty calculates achieved difficulty score from double-sha256 header hash LE
func HashToDifficulty(hashLE []byte) float64 {
	hashBE := ReverseBytes(hashLE)
	hashBigInt := new(big.Int).SetBytes(hashBE)
	if hashBigInt.Sign() == 0 {
		return math.MaxFloat64
	}
	diff1Float := new(big.Float).SetInt(Diff1Target)
	hashFloat := new(big.Float).SetInt(hashBigInt)
	resFloat := new(big.Float).Quo(diff1Float, hashFloat)
	res, _ := resFloat.Float64()
	return res
}

// CalculateMerkleRoot computes Merkle Root from Coinbase TxID (32-byte BE) and Merkle Branch Array
func CalculateMerkleRoot(coinbaseTxidBE []byte, merkleBranchHex []string) []byte {
	currentHash := coinbaseTxidBE

	for _, branchHex := range merkleBranchHex {
		branchBuf, err := hex.DecodeString(branchHex)
		if err != nil {
			continue
		}
		combined := append(currentHash, branchBuf...)
		currentHash = Sha256d(combined)
	}

	return currentHash
}

// BuildBlockHeader constructs an 80-byte Bitcoin Block Header Buffer
func BuildBlockHeader(version uint32, prevHashRawHex string, merkleRootBE []byte, nTimeHex, nBitsHex, nonceHex string) []byte {
	header := make([]byte, 80)

	// 1. Version (uint32 LE)
	binary.LittleEndian.PutUint32(header[0:4], version)

	// 2. PrevHash (32 bytes LE reversed from RPC BE)
	prevHashBE, _ := hex.DecodeString(prevHashRawHex)
	copy(header[4:36], ReverseBytes(prevHashBE))

	// 3. Merkle Root (32 bytes)
	copy(header[36:68], merkleRootBE)

	// 4. nTime (uint32 LE)
	nTimeVal, _ := strconv.ParseUint(nTimeHex, 16, 32)
	binary.LittleEndian.PutUint32(header[68:72], uint32(nTimeVal))

	// 5. nBits (uint32 LE)
	nBitsVal, _ := strconv.ParseUint(nBitsHex, 16, 32)
	binary.LittleEndian.PutUint32(header[72:76], uint32(nBitsVal))

	// 6. Nonce (uint32 LE)
	nonceVal, _ := strconv.ParseUint(nonceHex, 16, 32)
	binary.LittleEndian.PutUint32(header[76:80], uint32(nonceVal))

	return header
}

File: src/crypto/sha256.ts
import crypto from 'crypto';

/**
 * Standard Bitcoin SHA-256d (Double SHA-256)
 */
export function sha256d(data: Buffer): Buffer {
  const hash1 = crypto.createHash('sha256').update(data).digest();
  return crypto.createHash('sha256').update(hash1).digest();
}

/**
 * Reverse a Buffer (endianness swap)
 */
export function reverseBuffer(buffer: Buffer): Buffer {
  const res = Buffer.allocUnsafe(buffer.length);
  for (let i = 0; i < buffer.length; i++) {
    res[i] = buffer[buffer.length - 1 - i];
  }
  return res;
}

/**
 * Convert compact nBits representation to 256-bit BigInt target
 */
export function nbitsToTarget(nbitsHex: string): bigint {
  const nbits = parseInt(nbitsHex, 16);
  const exponent = nbits >> 24;
  const mantissa = BigInt(nbits & 0x00ffffff);
  if (exponent <= 3) {
    return mantissa >> BigInt(8 * (3 - exponent));
  }
  return mantissa << BigInt(8 * (exponent - 3));
}

/**
 * Convert compact nBits hex string to Network Difficulty number
 */
export function nbitsToDifficulty(nbitsHex: string): number {
  const target = nbitsToTarget(nbitsHex);
  if (target === 0n) return 0;
  return Number(DIFF1_TARGET) / Number(target);
}

/**
 * Difficulty 1 Target (0x00000000ffff0000000000000000000000000000000000000000000000000000)
 */
export const DIFF1_TARGET = 0x00000000ffff0000000000000000000000000000000000000000000000000000n;

/**
 * Convert numeric difficulty to 256-bit target BigInt
 */
export function difficultyToTarget(diff: number): bigint {
  if (diff <= 0) return DIFF1_TARGET;
  const target = BigInt(Math.floor(Number(DIFF1_TARGET) / diff));
  return target;
}

/**
 * Calculate achieved difficulty score from double-sha256 header hash Buffer (32 bytes)
 */
export function hashToDifficulty(hashLE: Buffer): number {
  const hashBE = reverseBuffer(hashLE);
  const hashBigInt = BigInt('0x' + hashBE.toString('hex'));
  if (hashBigInt === 0n) return Number.MAX_SAFE_INTEGER;
  return Number(DIFF1_TARGET) / Number(hashBigInt);
}

/**
 * Convert Buffer or hex to 256-bit BigInt for comparison
 */
export function bufferToBigIntBE(buffer: Buffer): bigint {
  return BigInt('0x' + buffer.toString('hex'));
}

/**
 * Calculate Merkle Root from Coinbase TxID (32-byte BE) and Merkle Branch Array (strings/Buffers)
 */
export function calculateMerkleRoot(coinbaseTxidBE: Buffer, merkleBranchHexArray: string[]): Buffer {
  let currentHash = coinbaseTxidBE;

  for (const branchHex of merkleBranchHexArray) {
    const branchBuf = Buffer.from(branchHex, 'hex');
    // Concatenate currentHash + branchBuf
    const combined = Buffer.concat([currentHash, branchBuf]);
    currentHash = sha256d(combined);
  }

  return currentHash;
}

/**
 * Reconstruct Header Version for AsicBoost (Version Rolling)
 * Formula: (baseVersion & ~mask) | (versionBits & mask)
 */
export function calculateAsicBoostVersion(
  baseVersionHex: string,
  versionBitsHex?: string,
  maskHex?: string
): number {
  const baseVersion = parseInt(baseVersionHex, 16);
  if (!versionBitsHex || !maskHex) {
    return baseVersion;
  }
  const versionBits = parseInt(versionBitsHex, 16);
  const mask = parseInt(maskHex, 16);

  // Bitwise composition
  const finalVersion = (baseVersion & ~mask) | (versionBits & mask);
  // Ensure unsigned 32-bit integer
  return finalVersion >>> 0;
}

/**
 * Construct 80-byte Bitcoin Block Header Buffer
 *
 * Bitcoin block header layout (80 bytes):
 *   [0..3]   version       (uint32 LE)
 *   [4..35]  prevHash      (32 bytes, internal byte order = LE of the hash)
 *   [36..67] merkleRoot    (32 bytes, internal byte order)
 *   [68..71] nTime         (uint32 LE)
 *   [72..75] nBits         (uint32 LE)
 *   [76..79] nonce         (uint32 LE)
 *
 * Stratum v1 miners send nTime and nonce as big-endian hex strings
 * representing uint32 values. We parse them to integers and write LE.
 *
 * prevHashRawHex is the RPC previousblockhash (big-endian display order).
 * For the header we need it in internal byte order (reversed = LE).
 */
export function buildBlockHeader(params: {
  version: number;
  prevHashRawHex: string;
  merkleRootBE: Buffer;
  nTimeHex: string;
  nBitsHex: string;
  nonceHex: string;
}): Buffer {
  const header = Buffer.alloc(80);

  // 1. Version (uint32 LE)
  header.writeUInt32LE(params.version >>> 0, 0);

  // 2. PrevHash: RPC gives BE display order, header needs LE (reversed)
  const prevHashBE = Buffer.from(params.prevHashRawHex, 'hex');
  reverseBuffer(prevHashBE).copy(header, 4);

  // 3. Merkle Root: sha256d output is already in internal byte order
  params.merkleRootBE.copy(header, 36);

  // 4. nTime: parse hex string as uint32 and write LE
  const nTimeVal = parseInt(params.nTimeHex, 16) >>> 0;
  header.writeUInt32LE(nTimeVal, 68);

  // 5. nBits: parse hex string as uint32 and write LE
  const nBitsVal = parseInt(params.nBitsHex, 16) >>> 0;
  header.writeUInt32LE(nBitsVal, 72);

  // 6. Nonce: parse hex string as uint32 and write LE
  const nonceVal = parseInt(params.nonceHex, 16) >>> 0;
  header.writeUInt32LE(nonceVal, 76);

  return header;
}
